File: eyan_helper/src/main/java/eu/eyan/util/swing/JProgressBarPlus.java
package eu.eyan.util.swing;

import static javax.swing.SwingUtilities.invokeLater;

import java.util.function.Consumer;

import javax.swing.JProgressBar;

@SuppressWarnings("serial")
public class JProgressBarPlus extends JProgressBar {

	private Consumer<Integer> percentChangedConsumer;
	private String format;

	public JProgressBarPlus(int min, int max, String format) {
		super(min, max);
		this.format = format;
		this.percentChangedConsumer = createPercentChangedConsumer();
	}

	public Consumer<Integer> createPercentChangedConsumer() {
		return percent -> invokeLater(() -> {
			this.setVisible(true);
			this.setString(String.format(format, percent));
			this.setValue(percent);
		});
	}

	public JProgressBarPlus percentChanged(int value) {
		if (this.getValue() != value && this.percentChangedConsumer != null) {
			this.percentChangedConsumer.accept(value);
		}
		return this;
	}

	public Runnable doneThenInvisible() {
		return () -> invokeLater(() -> this.setVisible(false));
	}

	public JProgressBarPlus setFormat(String format) {
		this.format = format;
		return this;
	}

	public JProgressBarPlus finished() {
		invokeLater(() -> {
			this.setString("Ready");
			this.setValue(100);
		});
		return this;
	}
}

File: eyan_helper/src/main/scala/eu/eyan/util/swing/JProgressBarPlus.scala
package eu.eyan.util.swing

import SwingPlus.invokeLater
import javax.swing.JProgressBar
import eu.eyan.util.swing.JComponentPlus.JComponentImplicit

object JProgressBarPlus {
  implicit class JProgressBarImplicit[TYPE <: JProgressBar](jProgressBar: TYPE) extends JComponentImplicit(jProgressBar) {
    //    addChangeListener(ChangeListener)
    //    removeChangeListener(ChangeListener)
    //    setBorderPainted(boolean)
    //    setIndeterminate(boolean)
    //    setMaximum(int)
    //    setMinimum(int)
    //    setModel(BoundedRangeModel)
    //    setOrientation(int)
    //    setString(String)
    //    setStringPainted(boolean)
    //    setUI(ProgressBarUI)
    //    setValue(int)
    //    updateUI()
  }
}
class JProgressBarPlus(min: Int = 0, max: Int = 100, var format: String = "%d%%", finishedText: String = "Ready") extends JProgressBar(min, max) {

  setNewValue(min)

  def setNewValue(percent: Int) = { setString(String.format(format, percent: Integer)); setValue(percent); this }

  def percentChanged(percent: Int) = { if (this.getValue() != percent) invokeLater { this.setVisible(true); setNewValue(percent) }; this }

  // refactor it.
  def doneThenInvisible = { finished; invokeLater(setVisible(false)); this }

  def finished = { invokeLater({ this.setString(finishedText); this.setValue(this.getMaximum()) }); this }

  def setFormat(newFormat: String) = {format = newFormat; this}
}